provider/cmd/pulumi-resource-pulumiservice/schema.json

```diff
@@ -323,9 +323,9 @@
           "description": "Shell commands to run before the Pulumi operation executes."
         },
         "environmentVariables": {
-          "type": "object",
-          "additionalProperties": {
-            "type": "string"
+          "type": "array",
+          "items": {
+            "$ref": "#/types/pulumiservice:index:EnvironmentVariable"
```

Comment:
This is a breaking change, and we should generally avoid making breaking changes to the highest extent possible. Is the reason for the type change only to avoid https://github.com/pulumi/pulumi-dotnet/issues/22?

Reply:
Yeah, breaking change is not great, I assumed building a backward compatible way would be overkill (and take like a week)
Two reasons - one is to avoid the bug, another is to allow twin values. Right now I'm keeping key plaintext, while value is stored using twin value: `valuePlain` and 'valueCipher'. Because my code knows the string `value`, I can add Plain or Cipher to it and find the property. However, if I had a map like before, I'd need two key value pairs, instead of two properties, and then I'd have to append keys with Plain or Cipher. And if I do that, I have no way of telling what the original key actually was ':D I could parse all and guess, but then there's always a chance someone calls their key "keyPlain" and that would screw it up further ':D

Reply:
I'm not sure I understand why this implementation detail needs to be surfaced to the user. Or why the map couldn't be a map of string to objects in the implementation rather than an array of objects like it is in the service.

I don't think a breaking change is acceptable here, let's chat and we can discuss the issues you're running into and what other strategies we have to get around it.